### ISSUE TITLE
# signs within SelectFormat

### ISSUE THREAD

**Issue by vogelsgesang** [SlexAxton/messageformat.js#105]
With the current implementation the messageformat string

```
{VAR1, plural,
  one {1}
  other {{VAR2, select,
    other {#}
  }}
}
```

is equivalent to

```
{VAR1, plural,
  one {1}
  other {{VAR2, select,
    other {{VAR2}}
  }}
}
```

I expected SelectFormat not to interfere with the `#` since [the standard for SelectFormat](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/SelectFormat.html) doesn't even mention the `#` sign.

If I am reading [the standard for PluralFormat](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/PluralFormat.html) correctly
```
An unquoted pound sign (#) in the selected sub-message
itself (i.e., outside of arguments nested in the sub-message)
is replaced [...]
```
the `#` in my string should simply yield a literal `#`.

Which behavior do you consider to be correct?

**Comment by eemeli** [comment]
You're right, our quoting doesn't match the standard. Here's the [official documention](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/MessageFormat.html)'s take on them (ignore the choiceStyle bit, we don't implement that):

> * messageText can contain quoted literal strings including syntax characters. A quoted literal string begins with an ASCII apostrophe and a syntax character (usually a {curly brace}) and continues until the next single apostrophe. A double ASCII apostrohpe inside or outside of a quoted string represents one literal apostrophe.
* Quotable syntax characters are the {curly braces} in all messageText parts, plus the '#' sign in a messageText immediately inside a pluralStyle, and the '|' symbol in a messageText immediately inside a choiceStyle.

And this is from our readme:

> messageformat.js tries to a good job of being tolerant of as much as possible, but some characters, like the ones used the actual MessageFormat spec itself, must be escaped to be a part of your string.

> For `{`, `}` and `#` (only inside of a select value) literals, just escape them with a backslash. (If you are in a JS string, you'll need to escape the escape backslash so it'll look like two).

At a minimum, we ought to consider `#` as a special character only where the standard so requires (there are even [examples](http://icu-project.org/apiref/icu4j/com/ibm/icu/text/SelectFormat.html) in the documentation that we throw an error for), but should we even consider switching from `"quoting \\{like\\} this"` to `"quoting '{like}' this"`? If the latter, the regex for quoted strings would be something like `/'[{}][^']*'/g` normally, but `/'[{}#][^']*'/g` at the top level within a pluralStyle -- except for the doubled single quotes; they'd need special handling.

**Comment by eemeli** [comment]
This is fixed in the 1.0.0 [pre-release version](/SlexAxton/messageformat.js/releases/tag/v1.0.0-rc.3).
